zcash_client_sqlite/src/lib.rs

```diff
@@ -445,6 +446,63 @@ impl<P: consensus::Parameters> WalletWrite for WalletDb<rusqlite::Connection, P>
         )
     }
 
+    fn insert_address_with_diversifier_index(
+        &mut self,
+        account: AccountId,
+        diversifier_index: DiversifierIndex,
+    ) -> Result<UnifiedAddress, SqliteClientError> {
+        self.transactionally(|wdb| {
+            let keys = wdb.get_unified_full_viewing_keys()?;
+            let ufvk = keys
+                .get(&account)
+                .ok_or(SqliteClientError::AccountUnknown(account))?;
+
+            let has_orchard = true;
```

Comment:
This can't be merged as-is, because this will cause Orchard receivers to begin being generated before the wallet has `orchard` support, which is a "the user can't see their received funds" risk. Instead, either use https://github.com/zcash/librustzcash/blob/main/zcash_client_sqlite/src/lib.rs#L110 or have the request passed in by the caller.

Reply:
That potentially presents a problem: if orchard cannot be guaranteed, and sapling may fail at a given diversifier index, that leaves me with just transparent in some cases. I can't construct a UA with just a transparent receiver. Yet the sqlite table I'm adding a row to has an `address` column which (at least so far) seems to only take UAs. Can I insert a transparent address there, or will code fall over?

Reply:
Looking at `get_next_available_address` I guess I don't see any insurance that T addresses aren't put into that table as it is, so maybe this isn't a new problem.

Reply:
Maybe I should just create a new table to track the transparent addresses. It's not in this PR, but I have a follow-up change in my fork that adds a column to the `addresses` table to record the last sync'd block for the transparent component of the address. If that isn't tolerable for this crate, then my creating a whole new table to store the data I need (and isn't wanted upstream) may be a better overall approach.

Reply:
If a Sapling receiver is requested and no valid diversifier exists at the specified index, that should result in an error being returned; we should _never_ generate an address that does not have all of the requested components.

Reply:
That doesn't really answer my question. I need to be able to store addresses for which no valid sapling receiver exists. Since I can't store orchard receivers in this table without the caller's permission, I have to either store transparent addresses in the `address` column, or find another table to store this data. Do you have a preference?

Reply:
Is the use case you're after here the "transparent address zero" problem, wherein older wallets may have generated a transparent address at index zero but there is no corresponding Sapling address, and so you can't generate a UA to correspond? 

That is a use case that we do need to handle, and I'd rather not special-case it, so we could potentially handle it by making the `address` column nullable, and using the `cached_transparent_receiver_address` for this purpose. However, for that, you'd need a different entrypoint, as at present you can't create UA requests that don't have a shielded component.

The alternative is that in ZIP 316 Revision 1, we could permit transparent-only UAs. Ultimately this might be what we do.

Reply:
No, this isn't directly related to the index 0 problem at all. This is simply about tracking which transparent addresses have been created/used so that sync knows which addresses to download transactions for.

Reply:
So up until now, you could obtain this from the transparent parts of UAs saved to the `addresses` table, plus the 0 address (which I had forgotten I added handling for in https://github.com/zcash/librustzcash/pull/685). Is the thing that is changing is that you intend to generate transparent addresses that are _not_ correlated to any UA? Sorry if I'm being obtuse, but I'm still having a bit of trouble understanding the use case.

Reply:
Yes. *All* my transparent addresses are generated completely independently of any UA. So I need to be able to represent all of them, even if their sapling counterpart is invalid.